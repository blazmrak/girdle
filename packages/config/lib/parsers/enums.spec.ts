import { initializeEnvironment } from '../index'
import { expectTypeOf } from 'expect-type'
import { extractError } from '../../tests/helpers/assertions'
import { Var } from './index'

enum Test {
  VALID = 'valid',
  OK = 'ok',
}

describe('enum', () => {
  it('parses with enum definition', () => {
    const env = { ENUM: 'valid', OD: 'ok' }

    const config = initializeEnvironment(
      {
        defaultEnum: Var.enum(Test).default(Test.VALID),
        od: Var.enum(Test).default(Test.VALID),
        optionalEnum: Var.enum(Test).optional(),
        enum: Var.enum(Test),
        namedEnum: Var.enum(Test).name('ENUM'),
      },
      { env },
    )

    expect(config).toMatchSnapshot()
    expectTypeOf(config).toMatchTypeOf<{
      defaultEnum: Test
      od: Test
      optionalEnum: Test | undefined
      enum: Test
      namedEnum: Test
    }>()
  })

  it('parses with array definition', () => {
    const env = { ENUM: 'valid', OD: 'ok' }

    const config = initializeEnvironment(
      {
        defaultEnum: Var.enum(['valid', 'ok'] as const).default('valid'),
        od: Var.enum(['valid', 'ok'] as const).default('valid'),
        optionalEnum: Var.enum(['valid', 'ok'] as const).optional(),
        enum: Var.enum(['valid', 'ok'] as const),
        namedEnum: Var.enum(['valid', 'ok'] as const).name('ENUM'),
      },
      { env },
    )

    expect(config).toMatchSnapshot()
    expectTypeOf(config).toMatchTypeOf<{
      defaultEnum: 'valid' | 'ok'
      od: 'valid' | 'ok'
      optionalEnum: 'valid' | 'ok' | undefined
      enum: 'valid' | 'ok'
      namedEnum: 'valid' | 'ok'
    }>()
  })

  it('fails if environment does not match the enum', () => {
    const env = { FAIL: 'invalid' }

    const e = extractError(() =>
      initializeEnvironment(
        {
          fail: Var.enum(Test),
        },
        { env },
      ),
    )

    expect(e).toMatchSnapshot()
  })
})
