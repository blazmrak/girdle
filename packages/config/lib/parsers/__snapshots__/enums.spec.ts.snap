// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enum parser fails if environment does not match the enum 1`] = `
EnvironmentInitializationFailed {
  "errors": [
    {
      "error": "Value of 'fail' (FAIL) should be one of [valid, ok]",
    },
  ],
}
`;

exports[`enum parser parses with array definition 1`] = `
{
  "defaultEnum": "valid",
  "enum": "valid",
  "env": "local",
  "namedEnum": "valid",
  "od": "ok",
  "optionalEnum": undefined,
}
`;

exports[`enum parser parses with enum definition 1`] = `
{
  "defaultEnum": "valid",
  "enum": "valid",
  "env": "local",
  "namedEnum": "valid",
  "od": "ok",
  "optionalEnum": undefined,
}
`;
